### PR TITLE
DOC: Add ignore-deprecate argument to validate_docstrings.py

### LINKED ISSUE #23631: CI: Add --ignore-deprecated to validate_docstrings.py
The script `scripts/validate_docstrings.py` returns the list of errors in the docstrings. We are using it to get the list of things to fix, and to check in the CI that no more mistakes are added, after the errors are fixed. See for example #23630.

As we don't really want to spend time fixing the docstrings that will be removed soon, it would be useful to add an argument to the script `--ignore-deprecated` that skips the functions/methods that are marked as deprecated (see `Docstring.deprecated`).

Things to note:
- When deprecated docstrings are skipped, they should not be displayed in the output of the script, and also not counted in the exit status of the script (so, the script should finish with exit status 0 if `--ignore-deprecated` is set and only deprecated functions are wrong).
- We should add at least one test to make sure the new argument works as expected.

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [X] closes #23631 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by datapythonista** [comment]
Thanks for the PR. Sorry the issue wasn't clear. Let me explain clear better...

`validate_docstrings.py` can be called for a single method, or for all them. If called for a single method, there is nothing to skip or do, as the most reasonable thing is to show all the errors for the requested docstring.

This is useful only when validating all the docstrings. That, instead of validating 100% of them, we can validate only the ones not deprecated (we don't care that much about errors in the docstrings of deprecated methods, as we are going to delete them in few weeks).

I think you did something different here, if I'm not wrong.

Btw, you should replace in the pull request description the `#xxxx` by the issue number you're fixing. Then, for every element in the checklist that applies, you should replace `[ ]` by `[X]`, so we know which items in the list you did.

Please let me know if you need additional info.

File: scripts/validate_docstrings.py
Comment (datapythonista) [inline scripts/validate_docstrings.py]: The argument name should probably be `--ignore-deprecated`, and just deprecated is not very clear

**Review by datapythonista — COMMENTED** [review]
That looks great now, sorry for the bad communication initially. Just added couple of comments about `argparse`.

File: scripts/validate_docstrings.py
Comment (charlesdong1991) [reply scripts/validate_docstrings.py]: yes, i also feel confused... because set it to true actually means we won't store it... 
changed...

File: scripts/validate_docstrings.py
Comment (charlesdong1991) [reply scripts/validate_docstrings.py]: nice one! the description looks much clear now!! thx!!!

**Review by datapythonista — APPROVED** [review]
lgtm, nice PR @charlesdong1991, thanks for the work on this

**Comment by charlesdong1991** [comment]
thank you @datapythonista  excited to learn from you!!

**Comment by charlesdong1991** [comment]
Hi, @datapythonista  sorry to bother you, i am struggling with debugging of pytest error, https://travis-ci.org/charlesdong1991/pandas/jobs/454418994, this fails every time. Since i added a new argument to the function validate_all, but it seems that this update is not synchronised in this pytest file since it shows it still accept only one argument... any idea how to fix it?

**Comment by datapythonista** [comment]
If I remember correctly, the exit status is the number of errors, not the number of docstrings failing. So, it should be 3, as one docstring generates 2 errors, the other 1, and the other is ignored because it's deprecated, so it will generate 0.

Does that make sense?

**Comment by charlesdong1991** [comment]
Hello, @datapythonista thank you very much for your quick response! Yes, i think exit status is the number of errors, however, in my case, i specified the error parameter to "ER01" which means it will only count the errors that are ER01, so in this case, only two will be chosen, because although the first docstring contains ER01, it's deprecated, so should not be counted. please correct me if i understand incorrectly.

**Comment by datapythonista** [comment]
I think the problem with the test is that the filtering is happening inside of `validate_all`, but you are actually testing `main` assuming that `validate_all` returns what's being mocked, which does not filter the deprecated docstrings. Does that make sense?

**Comment by charlesdong1991** [comment]
@datapythonista agree! i am thinking about the same, i also doubt if i should test the main... but i haven't found any test class or functions for validate_all? any suggestions on which is the best way to test? shall i create mock data as input and set deprecated to True, and then count the length of dictionary to see if the length is reduced accordingly to the deprecation?

**Comment by datapythonista** [comment]
I do think we have tests where we mock `validate_one` and test `validate_all`. Can you take a look? If we don't, we should create them anyway, but I think you should have them implemented so you can base the new one on those.

**Comment by charlesdong1991** [comment]
Hi, @datapythonista sorry to bother you again, i made new changes on pytest, and yesterday it seemed work and passed tests... however, today when i am trying to complete PR, i see some other errors that should not be caused by my change...
https://travis-ci.org/charlesdong1991/pandas/jobs/454724363, is it because the master branch has been updated? should i do git reset to catch the latest master before pushing?

**Comment by datapythonista** [comment]
Can't check the travis log from my phone, but it happens some times that things fail because of things unrelated to your PR.

Just leave your changes ready, and merge master in to your branch after some hours since the failure, and see if it's fixed.

Sometimes it can be that Anaconda is failing to download, a dependency released a new version that breaks something, problems in the Travis servers, something broken merged into master...

**Comment by charlesdong1991** [comment]
Hi, @datapythonista from my side, it looks like it passed all, 
https://travis-ci.org/charlesdong1991/pandas, however, the check on master still fails 
see https://travis-ci.org/pandas-dev/pandas/jobs/455383112, so can you take a look at what happens on master, and suggest a solution of how this can be fixed? thank you very much!

File: scripts/tests/test_validate_docstrings.py
Comment (datapythonista) [inline scripts/tests/test_validate_docstrings.py]: Move the curly bracket to the previous line

**Review by datapythonista — CHANGES_REQUESTED** [review]
Looks good, couple of formatting things

File: scripts/validate_docstrings.py
Comment (charlesdong1991) [reply scripts/validate_docstrings.py]: thanks for the correction!

File: scripts/validate_docstrings.py
Comment (charlesdong1991) [reply scripts/validate_docstrings.py]: removed!

File: scripts/tests/test_validate_docstrings.py
Comment (charlesdong1991) [reply scripts/tests/test_validate_docstrings.py]: did it!

File: scripts/tests/test_validate_docstrings.py
Comment (datapythonista) [inline scripts/tests/test_validate_docstrings.py]: Sorry I didn't see that before... `ignore_deprecated` expects a boolean value, I don't think we should pass None. And I see that it has a default value of `False` that is what we want. So, can you restore the original code? I think it should work, right?

File: scripts/tests/test_validate_docstrings.py
Comment (datapythonista) [inline scripts/tests/test_validate_docstrings.py]: same here, and in the rest of cases

File: scripts/tests/test_validate_docstrings.py
Comment (charlesdong1991) [reply scripts/tests/test_validate_docstrings.py]: ahh, i did this because i thought right now, `validate_all` takes in two arguments, and although `ignore_deprecated` has beed set to False by default, when we use lambda to create this mock data, we still need to assign a variable as a kind of placeholder, otherwise, there will be an error raised... can I still keep this, and remove None so that it looks like `lambda: prefix, ignore_deprecated` , please share your thoughts! And thanks for the quick response!

Comment (datapythonista) [reply scripts/tests/test_validate_docstrings.py]: oh, that's right, I checked it fast and didn't see those are the arguments of the mock function, not the parameters being passed. In that case set `ignore_deprecated=False`, so we have the same signature in the mock as in the mocked function.

Comment (charlesdong1991) [reply scripts/tests/test_validate_docstrings.py]: oh, yeah, changed to False! thx for the feedback!

**Review by datapythonista — APPROVED** [review]
lgtm, let's see if all tests pass now.

Thanks for the work on this @charlesdong1991

**Comment by charlesdong1991** [comment]
@datapythonista yeah, it passed! thx very much for your help!!

**Comment by datapythonista** [comment]
Tanks for the contribution @charlesdong1991. I'll let another maintainer double check the changes and merge them, but all good from my side.

**Comment by jreback** [comment]
@jorisvandenbossche or @TomAugspurger

**Review by jorisvandenbossche — APPROVED** [review]
Looks good to me!

**Pull request merged**
